File: hackerrank/easy/find-substring.cpp
// https://www.hackerrank.com/challenges/find-substring

#include <iostream>
#include <regex>
#include <utility>
#include <vector>

using namespace std;

int main() {
    size_t numSentences;
    cin >> numSentences >> ws;

    vector<string> sentences;
    sentences.reserve(numSentences);

    for (size_t i = 0; i < numSentences; ++i) {
        string sentence;
        getline(cin, sentence);
        sentences.push_back(move(sentence));
    }

    size_t numQueries;
    cin >> numQueries >> ws;

    for (size_t i = 0; i < numQueries; ++i) {
        string query;
        getline(cin, query);

        size_t match_count = 0;
        for (const auto &s : sentences) {
            regex regex("\\w" + query + "\\w");

            for (auto it = sregex_iterator(cbegin(s), cend(s), regex);
                 it != sregex_iterator();
                 ++it)
                ++match_count;
        }

        cout << match_count << "\n";
    }
}

File: hackerrank/easy/find-substring.java
// https://www.hackerrank.com/challenges/find-substring

import java.lang.*;
import java.util.*;
import java.util.regex.*;

public class Solution {
    public static void main(String[] args) {
        final Scanner s = new Scanner(System.in);

        final int numSentences = Integer.valueOf(s.nextLine());
        final List<String> sentences = new ArrayList<String>(numSentences);
        for (int i = 0; i < numSentences; ++i)
            sentences.add(s.nextLine());

        final int numQueries = Integer.valueOf(s.nextLine());
        for (int i = 0; i < numQueries; ++i) {
            final String query = s.nextLine();

            int matches = 0;
            for (String sentence : sentences) {
                final Matcher matcher = Pattern.compile("\\w" + query + "\\w").matcher(sentence);
                while (matcher.find())
                    ++matches;
            }

            System.out.println(matches);
        }
    }
}
